lorawan-encoding/src/certification.rs

```diff
@@ -0,0 +1,136 @@
+//! LoRaWAN Certification Protocol (TS009) command and payload handling
+use lorawan_macros::CommandHandler;
+
+use crate::maccommands::MacCommandIterator;
+use crate::maccommands::SerializableMacCommand;
+
+use crate::maccommands::Error;
+
+#[derive(Debug, PartialEq, CommandHandler)]
+#[cfg_attr(feature = "defmt-03", derive(defmt::Format))]
+pub enum DownlinkDUTCommand<'a> {
+    /// Request to reset the Microcontroller Unit
+    #[cmd(cid = 0x01, len = 0)]
+    DutResetReq(DutResetReqPayload),
+
+    /// Request to reset the LoRaWAN MAC layer and start issuing Join-Request messages
+    #[cmd(cid = 0x02, len = 0)]
+    DutJoinReq(DutJoinReqPayload),
+
+    /// Request to activate/deactivate Adaptive Data Rate (ADR)
+    #[cmd(cid = 0x04, len = 1)]
+    AdrBitChangeReq(AdrBitChangeReqPayload<'a>),
+
+    /// Change uplink periodicity to the provided value
+    #[cmd(cid = 0x06, len = 1)]
+    TxPeriodicityChangeReq(TxPeriodicityChangeReqPayload<'a>),
+
+    /// Send all subsequent uplinks of the specified type
+    // NB! Variable length payload
+    #[cmd(cid = 0x07)]
+    TxFramesCtrlReq(TxFramesCtrlReqPayload<'a>),
+
+    /// Request to send firmware version, LoRaWAN version, and regional parameters version
+    #[cmd(cid = 0x7f, len = 0)]
+    DutVersionsReq(DutVersionsReqPayload),
+}
+
+#[derive(Debug, PartialEq, CommandHandler)]
+#[cfg_attr(feature = "defmt-03", derive(defmt::Format))]
+pub enum UplinkDUTCommand<'a> {
+    #[cmd(cid = 0x7f, len = 12)]
+    DutVersionsAns(DutVersionsAnsPayload<'a>),
+}
+
+pub fn parse_downlink_certification_messages(
+    data: &[u8],
+) -> MacCommandIterator<'_, DownlinkDUTCommand<'_>> {
+    MacCommandIterator::new(data)
+}
```

Comment:
This API would lead to less confusion for the user
```suggestion
pub fn parse_downlink_certification_message(
    data: &[u8],
) -> Option<DownlinkDUTCommand<'_>> {
    let mut iter: MacCommandIterator<'_, DownlinkDUTCommand<'_>> = MacCommandIterator::new(data);
    iter.next()
}
```

Reply:
Ideally, the attribute macro could give us a function to try to parse just the next message so we don't make this degenerate use of the iterator, but this was easy to do at least (and we can fix it later without changing the API)

Reply:
Let's leave this for future. We use this construct for all the commands and it's extensively used in tests as well.